hy/cmdline.py

```diff
@@ -65,33 +68,48 @@ def __init__(self, spy=False, output_fn=None, locals=None,
             else:
                 self.output_fn = __builtins__[mangle(output_fn)]
 
-        code.InteractiveConsole.__init__(self, locals=locals,
-                                         filename=filename)
+        super(HyREPL, self).__init__(locals=locals, filename=filename)
 
         # Pre-mangle symbols for repl recent results: *1, *2, *3
         self._repl_results_symbols = [mangle("*{}".format(i + 1)) for i in range(3)]
         self.locals.update({sym: None for sym in self._repl_results_symbols})
 
-    def runsource(self, source, filename='<input>', symbol='single'):
-        global SIMPLE_TRACEBACKS
+    def error_handler(self, e):
+        self.locals[mangle("*e")] = e
+        self.showtraceback()
 
-        def error_handler(e, use_simple_traceback=False):
-            self.locals[mangle("*e")] = e
-            if use_simple_traceback:
-                print(e, file=sys.stderr)
-            else:
-                self.showtraceback()
+    def showsyntaxerror(self, filename=None):
+        if filename is None:
+            filename = self.filename
+
+        sys.last_type, sys.last_value, sys.last_traceback = sys.exc_info()
+
+        # Sadly, this method in Python 2.7 ignores an overridden `sys.excepthook`.
+        if sys.excepthook is sys.__excepthook__:
+            super(HyREPL, self).showsyntaxerror(filename=filename)
+        else:
+            sys.excepthook(sys.last_type, sys.last_value, sys.last_traceback)
+
+        self.locals[mangle("*e")] = sys.last_value
 
+    def showtraceback(self):
+        sys.last_type, sys.last_value, sys.last_traceback = sys.exc_info()
+
+        # Sadly, this method in Python 2.7 ignores an overridden `sys.excepthook`.
+        if sys.excepthook is sys.__excepthook__:
+            super(HyREPL, self).showtraceback()
+        else:
+            sys.excepthook(sys.last_type, sys.last_value, sys.last_traceback)
+
+        self.locals[mangle("*e")] = sys.last_value
+
+    def runsource(self, source, filename='<input>', symbol='single'):
```

Comment:
These two methods are very similar. Is it possible to factor out the common stuff?

Reply:
These two functions are part of the class spec and called implicitly elsewhere, so I can't get rid of them, but they can be simplified.